// typewriter/dist/commands/help.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@oclif/core");
const base_command_1 = require("../base-command");
class HelpCommand extends base_command_1.BaseCommand {
    async run() {
        const { argv } = await this.parse(HelpCommand);
        const Help = await (0, core_1.loadHelpClass)(this.config);
        const help = new Help(this.config, this.config.pjson.helpOptions);
        await help.showHelp(argv);
        this.segmentClient.helpCommand({
            properties: {
                rawCommand: this.rawCommand,
            },
        });
    }
}
exports.default = HelpCommand;
HelpCommand.description = "Display help for <%= config.bin %>.";
HelpCommand.flags = {
    ...base_command_1.BaseCommand.flags,
    "nested-commands": core_1.Flags.boolean({
        description: "Include all nested commands in the output.",
        char: "n",
    }),
};
HelpCommand.args = [
    {
        name: "command",
        required: false,
        description: "Command to show help for.",
    },
];
