// typewriter/dist/__tests__/ts-typedef-tests/tests/typescript-analytics-node-import-test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const analytics_node_1 = require("@segment/analytics-node");
const typescript_analytics_node_1 = tslib_1.__importDefault(require("../build/typescript-analytics-node"));
() => {
    // assert expects a node analytics instance
    typescript_analytics_node_1.default.setTypewriterOptions({
        analytics: new analytics_node_1.Analytics({ writeKey: "foo" }),
    });
    // @ts-expect-error - should require user ID OR anonId
    typescript_analytics_node_1.default.everyNullableOptionalType({
        properties: { "optional array with properties": [] },
    });
    // should require user ID OR anonId
    typescript_analytics_node_1.default.everyNullableOptionalType({
        userId: "foo",
        properties: { "optional array with properties": [] },
    });
    // should require user ID OR anonId
    typescript_analytics_node_1.default.everyNullableOptionalType({
        anonymousId: "foo",
        properties: { "optional array with properties": [] },
    });
    // @ts-expect-error - should require a properties object
    typescript_analytics_node_1.default.everyNullableOptionalType({
        anonymousId: "foo",
    });
    // assert empty properties
    typescript_analytics_node_1.default.everyNullableOptionalType({
        anonymousId: "foo",
        properties: {},
    });
    // assert context can be passed
    typescript_analytics_node_1.default.everyNullableOptionalType({
        anonymousId: "foo",
        context: {
            traits: {
                age: 99,
            },
            active: true,
        },
        properties: {},
    });
};
