// typewriter/dist/languages/javascript.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.javascript = void 0;
const typescript_1 = require("typescript");
const quicktype_utils_1 = require("./quicktype-utils");
const typescript_2 = require("./typescript");
class JavascriptGenerator {
    constructor() {
        this.supportedSDKs = typescript_2.typescript.supportedSDKs;
        this.defaultQuicktypeOptions = {
            'just-types': true,
        };
        this.generate = async (rules, context, options) => {
            const files = await typescript_2.typescript.generate(rules, context, options);
            const transpiledFiles = new Map();
            for (const [name, contents] of files.entries()) {
                const { outputText } = (0, typescript_1.transpileModule)(contents, {
                    compilerOptions: {
                        target: options.target,
                        module: options.module,
                        esModuleInterop: true,
                    },
                });
                transpiledFiles.set(name, outputText);
            }
            // TODO: Prettier
            return transpiledFiles;
        };
        const data = (0, quicktype_utils_1.getLanguageMetadata)('javascript');
        this.name = data.name;
        this.id = data.id;
        this.extension = data.extension;
        this.advancedOptions = [
            {
                type: 'list',
                name: 'target',
                message: 'Target',
                choices: [
                    { name: 'ES3', value: typescript_1.ScriptTarget.ES3 },
                    { name: 'ES5', value: typescript_1.ScriptTarget.ES5 },
                    { name: 'ES2015', value: typescript_1.ScriptTarget.ES2015 },
                    { name: 'ES2016', value: typescript_1.ScriptTarget.ES2016 },
                    { name: 'ES2017', value: typescript_1.ScriptTarget.ES2017 },
                    { name: 'ES2018', value: typescript_1.ScriptTarget.ES2018 },
                    { name: 'ES2019', value: typescript_1.ScriptTarget.ES2019 },
                    { name: 'ESNext', value: typescript_1.ScriptTarget.ESNext },
                    { name: 'Latest', value: typescript_1.ScriptTarget.Latest },
                ],
                default: typescript_1.ScriptTarget.ESNext,
            },
            {
                type: 'list',
                name: 'module',
                message: 'Module',
                choices: [
                    { name: 'CommonJS', value: typescript_1.ModuleKind.CommonJS },
                    { name: 'AMD', value: typescript_1.ModuleKind.AMD },
                    { name: 'UMD', value: typescript_1.ModuleKind.UMD },
                    { name: 'System', value: typescript_1.ModuleKind.System },
                    { name: 'ES2015', value: typescript_1.ModuleKind.ES2015 },
                    { name: 'ESNext', value: typescript_1.ModuleKind.ESNext },
                ],
                default: typescript_1.ModuleKind.ESNext,
            },
        ];
    }
}
exports.javascript = new JavascriptGenerator();
